File: scripts/oop-practice-sources/csharp/figure.cs
class Figure
{
    public string Name { get; }
    public string Color { get; }

    public Figure(string name, string color)
    {
        Name = name;
        Color = color;
    }

    public void Describe()
    {
        Console.WriteLine($"Фигура «{Name}», цвет: {Color}");
    }
}

class Circle : Figure
{
    public Circle(string color) : base("Круг", color) { }
}

class Square : Figure
{
    public Square(string color) : base("Квадрат", color) { }
}

class Program
{
    static void Main()
    {
        var circle = new Circle("красный");
        var square = new Square("синий");
        circle.Describe();
        square.Describe();
    }
}

File: scripts/oop-practice-sources/lua/figure.lua
local Figure = {}

function Figure:new(name, color)
    local obj = setmetatable({name = name, color = color}, {__index = self})
    return obj
end

function Figure:describe()
    print("Фигура «" .. self.name .. "», цвет: " .. self.color)
end

local Circle = {}
setmetatable(Circle, {__index = Figure})

function Circle:new(color)
    return Figure.new(self, "Круг", color)
end

local Square = {}
setmetatable(Square, {__index = Figure})

function Square:new(color)
    return Figure.new(self, "Квадрат", color)
end

local circle = Circle:new("красный")
local square = Square:new("синий")
circle:describe()
square:describe()
